ibis/expr/types/arrays.py

```diff
@@ -266,9 +266,16 @@ def repeat(self, n: int | ir.IntegerValue) -> ArrayValue:
 
     __mul__ = __rmul__ = repeat
 
-    def unnest(self) -> ir.Value:
+    def unnest(self,
+               offset: Optional[bool] = False
+        ) -> ir.Value:
```

Comment:
How does a user access the offset value? Wouldn't the result be something more complex than an array?

Perhaps this operation should be a new internal Ibis operation, like `UnnestWithOffset`.

Alternatively, the `dtype` property of the existing `Unnest` operation needs to be adjusted to return something like `array<struct<offset: int64, value: T>>`?

In any event, there's some API design work here.

Reply:
Thank you! I just noticed your comments here when I was creating #8356. Will follow up with your suggestion here!

Reply:
If the `dtype` property returns `struct` type as you suggested, it requires the transformed SQL selects a `struct` column too. But [explode_to_unnest](https://github.com/tobymao/sqlglot/blob/6cb985ae1346c1a912ed6f81be30310ee1c91dfa/sqlglot/transforms.py#L168) in `sqlglot` returns two columns instead. One tricky solution is creating a BQ custom `TRANSFORM` to modify the sqlglot expression tree. Do you have another suggestions?

```
In [12]: ibis.to_sql(table.select(["scalar_column", table.z.unnest(offset=True)]).z)
Out[12]: 
SELECT
  `t1`.`z`
FROM (
  SELECT
    `t0`.`scalar_column`,
    IF(pos = pos_2, `z`, NULL) AS `z`,
    IF(pos = pos_2, pos_2, NULL) AS pos_2
  FROM `bigframes-dev`.`ibis_gbq_testing`.`array_types` AS `t0`
  CROSS JOIN UNNEST(GENERATE_ARRAY(0, GREATEST(ARRAY_LENGTH(`t0`.`z`)) - 1)) AS pos
  CROSS JOIN UNNEST(`t0`.`z`) AS `z` WITH OFFSET AS pos_2
  WHERE
    pos = pos_2
    OR (
      pos > (
        ARRAY_LENGTH(`t0`.`z`) - 1
      ) AND pos_2 = (
        ARRAY_LENGTH(`t0`.`z`) - 1
      )
    )
) AS `t1`

In [13]: ibis.to_sql(table.select(["scalar_column", table.z.unnest(offset=True)]).pos_2)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[13], line 1
----> 1 ibis.to_sql(table.select(["scalar_column", table.z.unnest(offset=True)]).pos_2)

File ~/src/forkibis2/ibis/expr/types/relations.py:859, in Table.__getattr__(self, key)
    854     hint = common_typos[key]
    855     raise AttributeError(
    856         f"{type(self).__name__} object has no attribute {key!r}, did you mean {hint!r}"
    857     )
--> 859 raise AttributeError(f"'Table' object has no attribute {key!r}")

AttributeError: 'Table' object has no attribute 'pos_2'
```

Reply:
Updated the change to build an sge expression to combine the offset and value as a STRUCT node. But it is blocked by https://github.com/tobymao/sqlglot/issues/2978
Any thoughts are welcome!

Reply:
@cpcloud The `sqlglot` is not able to return a struct column combining `col` and `pos`, because `STRUCT(POSEXPODE(arr))` is not a `sparksql` grammar. Hence, I am thinking the ibis expression tree can construct two returned columns instead? If so, could you please give me some ideas how to achieve that?